pymeasure/adapters/protocol.py

```diff
@@ -0,0 +1,117 @@
+        """Generate the adapter and initialize internal buffers."""
+        super().__init__(preprocess_reply=preprocess_reply, **kwargs)
+        assert isinstance(comm_pairs, (list, tuple)), (
+            "Parameter comm_pairs has to be a list or tuple.")
+        for pair in comm_pairs:
+            if len(pair) != 2:
+                raise ValueError(f'Comm_pairs element {pair} does not have two elements!')
+        self._read_buffer = b""
+        self._write_buffer = b""
+        self.comm_pairs = comm_pairs
+        self._index = 0
+
+    def write(self, command):
+        """Compare the command with the expected one and fill the read."""
+        self.write_bytes(to_bytes(command))
+        assert self._write_buffer == b"", (
+            f"Written bytes '{self._write_buffer}' do not match expected "
+            f"'{self.comm_pairs[self._index][0]}'.")
+
+    def write_bytes(self, content):
+        """Write the bytes `content`. If a command is full, fill the read."""
+        self._write_buffer += content
+        p_write, p_read = self.comm_pairs[self._index]
+        if self._write_buffer == to_bytes(p_write):
+            assert self._read_buffer == b"", (
+                f"Unread response '{self._read_buffer}' present when writing. "
+                "Read the response; maybe a property's 'check_set_errors' is not accounted for?")
+            # Clear the write buffer
+            self._write_buffer = b""
+            self._read_buffer = to_bytes(p_read)
+            self._index += 1
+        # If _write_buffer does _not_ agree with p_write, this is not cause for
+        # concern, because you can in principle compose a message over several writes.
+        # It's not clear how relevant this is in real-world use, but it's analogous
+        # to the possibility to fetch a (binary) message over several reads.
+
+    def read(self):
+        """Return an already present or freshly fetched read buffer as a string."""
+        return (self.read_bytes(-1) + self.read_bytes(1)).decode("utf-8")
+
+    def read_bytes(self, count):
+        """Read `count` number of bytes."""
+        if self._read_buffer:
+            read = self._read_buffer[:count]
+            self._read_buffer = self._read_buffer[count:]
+            return read
+        else:
+            p_write, p_read = self.comm_pairs[self._index]
+            assert p_write is None, "Unexpected read without prior write."
+            self._read_buffer = to_bytes(p_read)[count:]
+            self._index += 1
+            return to_bytes(p_read)[:count]
+
+    # TODO: Harmonise ask being write+read (i.e., remove from VISAAdapter),
+    #   use protocol tests to confirm it works correctly
+    # TODO: Remove all now-unnecessary ask() implementations -
+    #   read and write implementations should have all (test this)
+    # TODO: Implement over-the-wire message-traffic logging here, first
+    # TODO: Check timing impact of non-firing logging messages
```

Comment:
Yeah, some of those should probably be de-scoped here -- maybe in a follow-up PR, so that we don't have to wait too long. I just did not have a better location at hand to record those todos so they don't get lost.

* 2 "Removal of ask" items - we are now in a position to confirm this does not break anything, especially if we first implement protocol test for instruments having an overridden `ask()`, so we can do this immediately after this PR.
* Logging: Indeed, that can wait for the restructuring. Maybe(?) we can do the `CMND`/`RESP` pair in Instrument.write/read that I mentioned [here](https://github.com/pymeasure/pymeasure/issues/627#issuecomment-1134310180) immediately, and postpone the `SEND`/`RECV` on Adapter/connection level for later?
* Logging timing: Also postpone until we implement the adapter logging. There is a remark in IK about this and a separate `debug` flag that I'm not sure is necessary, I'd like to confirm my hunch then.